source/adios2/ADIOSTypes.h

```diff
@@ -11,36 +11,169 @@
 #ifndef ADIOS2_ADIOSTYPES_H_
 #define ADIOS2_ADIOSTYPES_H_
 
+/// \cond EXCLUDE_FROM_DOXYGEN
 #include <complex>
 #include <cstddef>
 #include <cstdint>
 #include <limits>
+#include <map>
+#include <string>
 #include <type_traits>
+#include <vector>
+/// \endcond
 
 #include "adios2/ADIOSConfig.h"
 
 namespace adios
 {
 
-// Alias the fixed sized typed into the adios namespace to make sure we're
-// always using the right ones.
+/** Variable shape type identifier */
+enum class ShapeID
+{
+    GlobalValue, ///< single global value, common case
+    GlobalArray, ///< global (across MPI_Comm) array, common case
+    JoinedArray, ///< global array with a common (joinable) dimension
+    LocalValue,  ///< special case, local independent value
+    LocalArray   ///< special case, local independent array
+};
+
+/** Used to set IO class */
+enum class IOMode
+{
+    Independent, ///< all I/O operations are independent per rank
+    Collective   ///< expect collective I/O operations
+};
+
+enum class OpenMode
+{
+    Undefined,
```

Comment:
It might be a good idea to give these explicit values so the bit-wise operations make sense:
```
Undefined = 0b00000000,
Write     = 0b00000001,
Read      = 0b00000010,
Append    = 0b00000101,
ReadWrite = 0b00000011
```
This allows for things like:
```
// Write test, valid for Write, ReadWrite, and Append
if (val & OpenMode::Write)
```
and
```
// Read test, valid for Read and ReadWrite
if (val & OpenMode::Read)
```